### DISCUSSION TITLE
docker networks as a security boundary in swarm

### DISCUSSION THREAD

**Discussion by chrisbecke** [BretFisher/ama#203] [Q&A] [answered]
Docker networks are frequently touted as a security barrier in swarm but in practice don't really seem to work as advertized.

Given Docker Swarm is in an unclear place, and these talks are somewhat dated... In the Dogvscat demo mentioned in the DockerCon 2018 talk multiple networks are created to partition services in stacks from each other.

However, that example suffers from the same probelm Ive faced in my own deployments, almost everything connects to the traefik network, and is consequently enumerable to otherservices, no matter what stack, on the traefik network. Once you add additional networks to gather metrics like prometheus and or logging like fluentd, the network security boundaries are so much theory.

In practice this seems quite problematic over and above the security boundary being non existent... enumerating things like "tasks.api" might pick up all services called api, regardless of stack, attached to traefik, or some other common network and have undesirable consequences.

Ive never seen any discussion, or potential solutions for this issue.

**Comment by BretFisher** [comment, marked as answer]
You should only attach public-facing containers to the proxy network.  All other containers like db's, queues, workers, would be on a different docker network. The public-facing containers like web and API would attach to both networks. In this way, it's similar to how you would have a DMZ design in a physical network setup.

**Comment by chrisbecke** [comment]
Even then, there should be a lot more care taken to ensure that the services that do sit in multiple networks do not bind to all network interfaces. But swarm gives no token that can be used in a stack file that represents the ip of a service on a network (which most images require as a command: parameter), and figuuring it out from interface names like eth1, eth2 using an entrypoint script is hard because docker assigns networks in hash table order, and there doesnt seem to be a general way to get the local host ip for network with index "n" anyway.

**Comment by BretFisher** [comment]
Using fixed IP's in cloud native (containers) is generally discouraged and I've been able to avoid it in every client deployment over the last 5 years. I imagine there's a way to do what you need with DNS.
